main.cpp

```diff
@@ -104,144 +100,91 @@ struct DoubleType
         value = nullptr;
     }
 
-    DoubleType& add(double d) 
-    { 
-        *value += d; 
-        return *this; 
-    }
-    DoubleType& subtract(double d) 
-    { 
-        *value -= d; 
-        return *this; 
-    }
-    DoubleType& multiply(double d) 
-    { 
-        *value *= d; 
-        return *this; 
-    }
-    DoubleType& divide(double d) 
-    { 
-        if(d == 0) { std::cout << "Warning: dividing by 0 results in undefined value!" << std::endl; }
-        *value /= d; 
-        return *this; 
-    }
-
-    // add/subtract/multiply/divide IntType 
+    DoubleType& add(double d);
+    DoubleType& subtract(double d);
+    DoubleType& multiply(double d);
+    DoubleType& divide(double d);
 
-    DoubleType& add(const IntType& it);
-    DoubleType& subtract(const IntType& it);
-    DoubleType& multiply(const IntType& it);
-    DoubleType& divide(const IntType& it);
+    operator double() { if(value) return *value; }
 
-    // add/subtract/multiply/divide FloatType 
+private:
 
-    DoubleType& add(const FloatType& it);
-    DoubleType& subtract(const FloatType& it);
-    DoubleType& multiply(const FloatType& it);
-    DoubleType& divide(const FloatType& it);
+    double* value = nullptr;
 };
 
-struct IntType
+DoubleType& DoubleType::add(double d)
 {
-    int* value = new int;
+    *value += d;
+    return *this;
+}
+
+DoubleType& DoubleType::subtract(double d)
+{
+    *value -= d;
+    return *this;
+}
+
+DoubleType& DoubleType::multiply(double d)
+{
+    *value *= d;
+    return *this;
+}
+
+DoubleType& DoubleType::divide(double d)
+{
+    if(d == 0.0) { std::cout << "Warning: dividing by 0 results in undefined value!" << std::endl; }
+    *value /= d;
+    return *this;
+}
 
+struct IntType
+{
     IntType(int i) : value(new int(i)) { }
     ~IntType()
     {
         delete value;
         value = nullptr;
     }
 
-    IntType& add(int i) 
-    { 
-        *value += i; 
-        return *this; 
-    }
-    IntType& subtract(int i) 
-    { 
-        *value -= i;
-        return *this; 
-    }
-    IntType& multiply(int i) 
-    { 
-        *value *= i; 
-        return *this; 
-    }
-    IntType& divide(int i) 
-    { 
-        if(i == 0) 
-        { 
-            std::cout << "Warning: dividing by 0 results in undefined value!" << std::endl;
-        } 
-        else 
-        {
-            *value /= i; 
-        }
-        return *this;
-    }
-
-    // add/subtract/multiply/divide DoubleType 
+    IntType& add(int i);
+    IntType& subtract(int i);
+    IntType& multiply(int i);
+    IntType& divide(int i);
 
-    IntType& add(const DoubleType& it);
-    IntType& subtract(const DoubleType& it);
-    IntType& multiply(const DoubleType& it);
-    IntType& divide(const DoubleType& it);
+    operator int() { if(value) return *value; }
```

Comment:
you know `value` is not a `nullptr` because of how the class is constructed. 
- [ ] remove the check to clear the warning.